### ISSUE TITLE
 Training hangs with sample_packing for specific datasets [multi-GPU]

### ISSUE THREAD

**Issue by jphme** [axolotl-ai-cloud/axolotl#494]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Training should run normally.

### Current behaviour

With specific datasets and a multi-GPU setup (2*RTX 4090s in my case), training hangs (after the first training step or after the fist epoch) until a timeout error occurs.

The bug only happens:

- with 2 (or more?) GPUs
- with specific datasets (the average/maximum number of tokens per example seems to be important)
- with `sequence_len: 4096` (2048 runs fine)
- with `sample_packing: true`

With `micro_batch_size: 1` instead of 2 (which is also used for `sample_packing_seq_len_mulitpler`!), training starts fine but hangs after epoch 1 has finished.

When training hangs, both GPUs are stuck below 1/3 TWP (~90W in my case) and VRAM is still allocated.
P2P settings like `NCCL_P2P_DISABLE=1` don't change anything. 

### Steps to reproduce

Steps to reproduce (with config from [Llama-2/examples/lora.yml](https://github.com/OpenAccess-AI-Collective/axolotl/blob/main/examples/llama-2/lora.yml) below.
Minimal example dataset to reproduce the issue is here [jphme/test_ds_2gpu_axolotl_bug](https://huggingface.co/datasets/jphme/test_ds_2gpu_axolotl_bug).

```bash
#works fine
accelerate launch scripts/finetune.py examples/llama-2/lora.yml

#hangs with 2gpus, nothing except datasegt changed
accelerate launch scripts/finetune.py examples/llama-2/lora_w_dataset_above.yml
```

Reducing sequence_len to 2048 or disabling sample_packing prevents the hanging; when setting `micro_batch_size: 1` training runs fine until epoch 1 has finished and then hangs. 

### Possible solution

I suspect that the bug has something todo with the `sample_packing` implementation and is only triggered for examples with many tokens:

Numer of tokens of the original Llama-2 lora.yml dataset `mhenrichsen/alpaca_2k_test` which runs fine:
```bash
count    2000.000000
mean      217.408000
std       146.209145
min        49.000000
25%        95.000000
50%       167.000000
75%       313.000000
max       762.000000
dtype: float64
```

Token length of my custom dataset:
```
count      25.000000
mean     1805.000000
std       603.574629
min       636.000000
25%      1274.000000
50%      1885.000000
75%      2184.000000
max      2991.000000
dtype: float64
```

As training hangs only for long sequence lengths, multiple GPUs and dataset with examples with many tokens, it should be possible to identify the root cause; unfortunately im out of my depth with the sample_packing code. 

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.9.17 (via Dockerfile)

### axolotl branch-commit

main/f144e98a3259fafcb6d359147b756816333a227e

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

cc @winglian

**Comment by tmm1** [comment]
Can you use pyspy to see where it's stuck?

**Comment by jphme** [comment]
How to do that? Never used pyspy ;-) .

But I got along further now, it seems to be some weird NCCL issue with distributed dataloaders.
With NCCL Async error handling and Accelerate debug i got the following output after the timeout:

```
ProcessGroupNCCL.cpp:828] [Rank 1] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=42, OpType=BROADCAST, Timeout(ms)=1800000) ran for 1809852 milliseconds before timing out.
[E ProcessGroupNCCL.cpp:455] Some NCCL operations have failed or timed out. Due to the asynchronous nature of CUDA kernels, subsequent GPU operations might run on corrupted/incomplete data.
[E ProcessGroupNCCL.cpp:460] To avoid data inconsistency, we are taking the entire process down.
terminate called after throwing an instance of 'std::runtime_error'
  what():  [Rank 1] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=42, OpType=BROADCAST, Timeout(ms)=1800000) ran for 1809852 milliseconds before timing out.
```

Going throgh almost all [NCCL settings](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html#nccl-p2p-level), I found that the setting
`-e NCCL_BUFFSIZE=536870912` (=512 MB, default is 4 MB) fixes the problenm and training runs through, even with  `micro_batch_size: 2` .

Do you know some source discussing these settings and their tradeoffs (do I have 512MB less VRAM per GPU now?) ?- I found basically nothing with Google and it was pure trial&error.....

EDIT: Now training completes, but I get some strange AssertionError afterwards (will rebuild container to be sure that this has no other sources):

```bash
{'loss': 5.7284, 'learning_rate': 4.8943483704846475e-06, 'epoch': 7.6}
{'loss': 5.7308, 'learning_rate': 0.0, 'epoch': 8.0}
100%|███████████████████████████████████████████████████████████████████████| 20/20 [04:40<00:00, 14.45s/it]/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer_pt_utils.py:289: FutureWarning: SequentialDistributedSampler is deprecated and will be removed in v5 of Transformers.
  warnings.warn(
/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer_pt_utils.py:289: FutureWarning: SequentialDistributedSampler is deprecated and will be removed in v5 of Transformers.
  warnings.warn(
[2023-08-28 16:08:33,192] [INFO] [axolotl.utils.dataloader._len_est:262] [PID:98] packing_efficiency_estimate: 0.86 total_num_tokens per device: 979
[2023-08-28 16:08:33,192] [INFO] [axolotl.utils.dataloader._len_est:262] [PID:99] packing_efficiency_estimate: 0.86 total_num_tokens per device: 979
[2023-08-28 16:08:33,197] [INFO] [axolotl.utils.dataloader.generate_batches:181] [PID:98] generating packed batches
[2023-08-28 16:08:33,197] [INFO] [axolotl.utils.dataloader.generate_batches:181] [PID:99] generating packed batches
Traceback (most recent call last):
  File "/workspace/axolotl/scripts/finetune.py", line 326, in <module>
    fire.Fire(train)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 475, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
Traceback (most recent call last):
  File "/workspace/axolotl/scripts/finetune.py", line 326, in <module>
    component = fn(*varargs, **kwargs)
  File "/workspace/axolotl/scripts/finetune.py", line 303, in train
    fire.Fire(train)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 141, in Fire
    trainer.train(resume_from_checkpoint=resume_from_checkpoint)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 1555, in train
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 475, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
    return inner_training_loop(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 1929, in _inner_training_loop
    component = fn(*varargs, **kwargs)
  File "/workspace/axolotl/scripts/finetune.py", line 303, in train
    trainer.train(resume_from_checkpoint=resume_from_checkpoint)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 1555, in train
    self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 2256, in _maybe_log_save_evaluate
    return inner_training_loop(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 1929, in _inner_training_loop
    metrics = self.evaluate(ignore_keys=ignore_keys_for_eval)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 2972, in evaluate
    self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 2256, in _maybe_log_save_evaluate
    metrics = self.evaluate(ignore_keys=ignore_keys_for_eval)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 2972, in evaluate
    output = eval_loop(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 3151, in evaluation_loop
        output = eval_loop(for step, inputs in enumerate(dataloader):

  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 3151, in evaluation_loop
  File "/workspace/axolotl/src/axolotl/utils/dataloader.py", line 214, in __iter__
    all_batches, _ = self.generate_batches(set_stats=True)
  File "/workspace/axolotl/src/axolotl/utils/dataloader.py", line 183, in generate_batches
    indices = [idx for idx in self.sampler]
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer_pt_utils.py", line 318, in __iter__
    assert (
AssertionError: Indices length 2 and total size 4 mismatched
    for step, inputs in enumerate(dataloader):
  File "/workspace/axolotl/src/axolotl/utils/dataloader.py", line 214, in __iter__
    all_batches, _ = self.generate_batches(set_stats=True)
  File "/workspace/axolotl/src/axolotl/utils/dataloader.py", line 183, in generate_batches
    indices = [idx for idx in self.sampler]
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer_pt_utils.py", line 318, in __iter__
    assert (
AssertionError: Indices length 2 and total size 4 mismatched
```

**Comment by jphme** [comment]
For further reference, the following Environment vars help to debug multi-gpu /hanging issues without stacktraces:
```
NCCL_ASYNC_ERROR_HANDLING=1
TORCH_SHOW_CPP_STACKTRACES=1
ACCELERATE_DEBUG_MODE="1" (or accelerate launch --debug)
```

Will close this in a few days if I find nothing else - the issue still seems not be solved (and I have a few ideas how to fix this with accelerate on the code side), but the main problems are apparently not Axolotl-related ...

**Comment by jphme** [comment]
Wow, so after 3 days debugging I finally feel like I understand the problem and am close to a solution - what follows will be a long text, partially still a WIP. 

**TL;DR: decrease `micro_batch_size` and `gradient_accumulation_steps`, increase `ddp_bucket_cap_mb`,  `NCCL_BUFFSIZE` and set `ddp_broadcast_buffers=False` if possible (or ignore selected buffers) - then hope for the best or new Drivers...**

# Main Problem: Training hangs / is stuck with more than one GPU (P2P disabled)

The timing/occurence changes, depending on dataset, batch_size and gradien_accumulation_steps and other parameters. Steps to reproduce (for my setup -> 2* RTX 4090s + Ryzen 7950) see at the end.
In my case, only occurs when using the `sample_packing` option in Axolotl (which was recently introduced).

First of all, the core problem is not Axolotl-specific, but seems to be a Buffer Problem with NCCL when P2P is disabled (like with two RTX 4090s), especially with AMD CPUs (in my case a Ryzen x7950). See similar issues e.g. [here](https://github.com/NVIDIA/nccl/issues/814), [here](https://github.com/huggingface/accelerate/issues/1311) or [here](https://www.pugetsystems.com/labs/hpc/problems-with-rtx4090-multigpu-and-amd-vs-intel-vs-rtx6000ada-or-rtx3090/). The newer NVIDA drivers should have fixed this, but some kind of buffer problem is still there. 

The hanging problems always occurs during the gradient accumulation step `self.accelerator.accumulate` in [this line](`https://github.com/huggingface/transformers/blob/716bb2e3910fd4872064c55b0d8bc3dad754d129/src/transformers/trainer.py#L1835C1-L1837C69) of the trainer, but without any stacktrace (and non-descript NCCL debug messages) I did not see a clear bug here and various Trainer/Dataloader parameters did not change anything. 

Most of the time the problem occured at start of the training or after the end of an epoch.

## NCCL
Due to the randomness I first tried various NCCL fixes (the nccl_tests ran fine with P2P disabled). See [here](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html) for all environment options.

I noticed that increasing the NCCL Buffer `-e NCCL_BUFFSIZE=536870912` (acutally I found almost no discussion/documentation on this setting and whether it affects VRAM or only system ram),  helped in some cases (albeit not reliably). 
I also played with som other NCCL settings like NCCL_MIN_NCHANNELS, NCCL_NTHREADS, NCCL_CHECKS_DISABLE but didn't notice any improvements.

## DDP
Next I explored various accelerate options and found the [DDP Documentation](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html), where `ddp_bucket_cap_mb` and `ddp_broadcast_buffers` and be set via Trainer kwargs. `ddp_bucket_cap_mb` helped a little bit in some cases (as NCCL_BUFFER_SIZE) but not reliably; however `ddp_broadcast_buffers` fixed the problem (for the time being?).

Then I also found some discussions of buffer problems with DDP (e.g. [here](https://github.com/pytorch/pytorch/issues/99652)) and there exists a better solution to selectively disable buffer/module sync for problematic buffers (which seems to be the case with the irregular-sized `sample_backing` batches?), see [here](https://github.com/pytorch/pytorch/issues/43690): [_set_params_and_buffers_to_ingore_for_model](https://github.com/pytorch/pytorch/blob/7ec439206853ec1a7ba74cfeeaa15d45d30475cc/torch/nn/parallel/distributed.py#L2149).

I still don't know which buffer exactly causes the problem in this case (and suspect it has something to do with the changing efficience/length estimates with `sample_packing`?) but unfortunately don't have more time to debug an am already out of my depth here. @tmm1 maybe you know immediately what to change (otoh its probably not a priority if I am the only one affected here). 

I will prepare a PR to make the ddp vars optionally available via the axolotl config file -> see #515

## Reproduction
I created a dataset with random strings of variable (and large in comparison to the context window) token length, because this helps replicating the issue.
`accelerate launch --multi_gpu --mixed_precision=bf16 axolotl/scripts/finetune.py lora_2gpu_hanging_bug.yml `

lora_2gpu_hanging_bug.yml:

```
base_model: meta-llama/Llama-2-7b-hf
base_model_config: meta-llama/Llama-2-7b-hf
model_type: LlamaForCausalLM
tokenizer_type: LlamaTokenizer
is_llama_derived_model: true

load_in_8bit: true
load_in_4bit: false
strict: false

datasets:
  - path: jphme/test_ds_2gpu_axolotl_bug
  #mhenrichsen/alpaca_2k_test
    type: alpaca
dataset_prepared_path: last_run_prepared
val_set_size: 0.01
output_dir: ./lora-out

sequence_len: 4096
sample_packing: true

adapter: lora
lora_model_dir:
lora_r: 32
lora_alpha: 16
lora_dropout: 0.05
lora_target_linear: true
lora_fan_in_fan_out:

wandb_project:
wandb_entity:
wandb_watch:
wandb_run_id:
wandb_log_model:

gradient_accumulation_steps: 4
micro_batch_size: 2
num_epochs: 10
optimizer: adamw_bnb_8bit
lr_scheduler: cosine
learning_rate: 0.0002

train_on_inputs: false
group_by_length: false
bf16: true
fp16: false
tf32: false

gradient_checkpointing: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 10
eval_steps: 20
save_steps:
debug:
deepspeed:
weight_decay: 0.0
fsdp:
fsdp_config:
special_tokens:
  bos_token: "<s>"
  eos_token: "</s>"
  unk_token: "<unk>"
 ```
 
Error happens with cu118 (deps like in Axolotl rep), as well as with cu121 and newer dependency versions; I am running Nvidia driver 535.86.05 but same error happened with 525.x .

pip freeze:
```
axolotl @ git+https://github.com/OpenAccess-AI-Collective/axolotl.git@98bf76e236ee7450cdfde506d9f409e5e28cb2a2
transformers==4.31.0
torch==2.1.0a0+b5021ba
accelerate @ git+https://github.com/huggingface/accelerate.git@a16b843a1b8626d571e62f7301e3a478be2c04b4
```

**Comment by dongxiaolong** [comment]
I had the same problem when I used 12*4090 gpu and platypus dataset.Do you have any new developments to deal with this?

**Comment by geronimi73** [comment]
same on 2x3090 w/ ReLoRA, sample_packing true or false, depends on bs if it works or not

**Comment by jphme** [comment]
> I had the same problem when I used 12*4090 gpu and platypus dataset.Do you have any new developments to deal with this?

See https://github.com/OpenAccess-AI-Collective/axolotl/pull/517 (and comments), seems to be a problem with divergent dataloader buffers on non-p2p setups, with that PR it should work (but the PR has other disadvantages, namely that the dataset has to be loaded fully on each GPU).

**Comment by veya2ztn** [comment]
can you just quickly try accelerate==0.20.0 ? 
In my case, accelerate == 0.22.0 has such problem but downgrade to 0.20.0 just fine.
I wonder my problem is same as yours...

**Comment by jphme** [comment]
Just wanted to comment quickly for further reference that #531 doesnt fix this issue, see #592 for new details/work on this.

**Comment by Foreist** [comment]
Setting up NCCL can eliminate messy stuttering on multi-GPUs.
